[Ngu Yi Yang] iP

nguyiyang opened the pull request:
# Lania
> Your **personal** chatbot to handle your tasks

Forget about remembering your own tasks, _Lania_ does it for you. It's
* easy to use
* fast
* friendly :blush:

Start using _Lania_ by doing the following,
1. download it from https://github.com/nguyiyang/ip/releases/tag/v0.1.
2. double-click it.
3. add and manage your tasks.
4. **enjoy**!

Features include:
- [x] Managing tasks
- [x] Finding tasks
- [ ] Graphical User Interface(coming soon)

Here's a short snippet of the Java program if you're interested! The `main` method:
```javascript
public static void main(String[] args) {
        Lania lania = new Lania("data/lania.txt");
        lania.run();
    }
}
```

Commit:
Add Gradle support

nguyiyang committed:
add greeting, echo and exit functionality to Duke.java

nguyiyang committed:
add update method in Lania.java to store user input

nguyiyang committed:
add list method to Lania.java to display tasks

nguyiyang committed:
add Task.java

nguyiyang committed:
replace string arrray with task array in Lania.java

nguyiyang committed:
add complete method and handling in Task.java

nguyiyang committed:
add string representation for task

nguyiyang committed:
change printing of task to use Task toString method

nguyiyang committed:
add Deadline.java Event.java Todo.java

nguyiyang committed:
change update method in Lania.java to handle different tasks

nguyiyang committed:
add number of tasks display for update method in Lania.java

nguyiyang committed:
Add testcases for Lania.java

nguyiyang committed:
Add LaniaException.java

nguyiyang committed:
Add LaniaEmptyDescriptionException.java

Added new exception that is a subclass of the LaniaException class to handle more specific errors.

nguyiyang committed:
Add ability to handle exceptions in Lania.java

nguyiyang committed:
Add testcase for user input errors

nguyiyang committed:
Change task array to use ArrayList

nguyiyang committed:
Add support for deleting tasks

nguyiyang committed:
Add run method and change all methods to non static

nguyiyang committed:
Add method to read file data at start of program

nguyiyang committed:
Add /data/lania.txt to .gitignore

nguyiyang committed:
Add write file method when task list is updated

nguyiyang committed:
Change format of string written in lana.txt

nguyiyang committed:
Update print file to also read data into array

nguyiyang committed:
Add /data/lania.txt to .gitignore

nguyiyang committed:
Update tasks to handle date and time formats

nguyiyang committed:
Add exception handling for invalid date/time format

nguyiyang committed:
Merge branch 'branch-Level-7'

nguyiyang committed:
Merge branch 'branch-Level-8'

nguyiyang committed:
Update text-ui-test testcases

nguyiyang committed:
Add Ui.java

nguyiyang committed:
Remove echo method in Lania.java

nguyiyang committed:
Add Storage.java

nguyiyang committed:
Refactor empty file handling to Storage.java

nguyiyang committed:
Add Parser.java

nguyiyang committed:
Add TaskList.java and refactor Task ArrayList

nguyiyang committed:
Add all class in lania package

nguyiyang committed:
Add task-related classes to task package

nguyiyang committed:
Add exception-related classes to exception package

nguyiyang committed:
Fix bug in constructor for LaniaException.java

nguyiyang committed:
Change constructor for TaskList.java

nguyiyang committed:
Change time format for Event.java and Deadline.java

nguyiyang committed:
Add JUnit tests in LaniaText.java

nguyiyang committed:
Move LaniaTest.java to test folder

nguyiyang committed:
Package app as JAR file

nguyiyang committed:
Add JavaDoc comments

nguyiyang committed:
Update code to comply with coding standard

nguyiyang committed:
Add find method in TaskList.java

nguyiyang committed:
Merge branch 'branch-A-JavaDoc'

nguyiyang committed:
Resolve merge conflict from branch-A-CodingStandards

nguyiyang committed:
Update code according to coding standards

nguyiyang committed:
Add Command class and subclasses

nguyiyang committed:
Merge branch 'add-gradle-support'

nguyiyang committed:
Change classname in build.gradle

nguyiyang committed:
Add missing handler for bye command in Parser

nguyiyang committed:
Refactor command classes into command package

nguyiyang committed:
Add config folder for checkstyle

nguyiyang committed:
Add JavaDoc to command classes

nguyiyang committed:
Merge branch 'branch-A-CheckStyle'

KelvinSoo reviewed (commented):
The code looks impressive! I can't find any coding violations.

nguyiyang committed:
Change implementation of commands

nguyiyang committed:
Add JavaFX

nguyiyang committed:
Update GUI style

nguyiyang committed:
Refactor Gui

nguyiyang committed:
Add DialogBox FXML

nguyiyang committed:
Merge branch 'branch-Level-10'

DonnyS57 commented on src/main/java/lania/Parser.java:
```diff
@@ -0,0 +1,128 @@
+     * @param fullCommand The input given by user.
+     * @return The type of command to be executed.
+     */
+    public Command parse(String fullCommand) throws LaniaException {
```
This code is fine on its own but perhaps a switch statement can help to increase the readability of the code?

nguyiyang replied:
Thank you for pointing that out, I agree with you and will do so to improve the readability.

DonnyS57 reviewed (commented):
Aside from the single method that I commented, everything else looks good! All the methods are very clear on their purpose and very readable! Keep up the amazing work 👍

dannylty commented on src/main/java/lania/Parser.java:
```diff
@@ -0,0 +1,128 @@
+        } else if (firstCommand.equals("delete")) {
+            return new DeleteCommand(getIndex(fullCommand));
+        } else {
+            if (firstCommand.equals("todo")) {
```
I understand that these lines below refer to commands that create new tasks but would it be better to either keep it at the same indentation or abstract this out into another method to create tasks?

nguyiyang replied:
Thank you for pointing that out, I think keeping them at the same indentation would be great.

dannylty commented on src/main/java/lania/Parser.java:
```diff
@@ -0,0 +1,128 @@
+     * @return The description of the given task.
+     * @throws LaniaEmptyDescriptionException If the task description is not given.
+     */
+    public String parseTaskDescription(String command) throws LaniaEmptyDescriptionException {
```
Would it be good to standardize the word "Description" in the method names given that the below two methods for Deadline and Event does not contain it?

nguyiyang replied:
Thank you for pointing that out, I think that would make the methods more consistent and easily understood.

dannylty commented on src/main/java/lania/Storage.java:
```diff
@@ -0,0 +1,97 @@
+                String next = s.nextLine();
+                String[] split = next.split("\\|", 4);
+                Task t;
+                if (split[0].equals("T")) {
```
I didn't like the raw logic being in the 4th indentation level here. Consider packaging it into a separate method?

nguyiyang replied:
Thank you for the advice, abstracting the code into a method would indeed make it cleaner. Will do so!

dannylty commented on src/main/java/lania/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * Another string representation of the deadline object.
```
Maybe include the differences between the two convert-to-String methods. Applies for the below classes too.

dannylty reviewed (commented):
I liked how every method and class has a description/comments. Good job!

nguyiyang committed:
Fix exit command for gui

nguyiyang committed:
Add assertions to program

The program lacks assertions to ensure various assumptions.

Some important assumptions should hold at various points in the code to
indicate any mistakes in the code.

Let's,
* add assertions to check that the task list loaded is not null.
* add assertions to check tasks in the task list are properply updated.
* add assertions check that the storage written to contains text in the
  proper format.

Using assertions this way can ensure that we have a task list to work
with when the program runs, and verify that the storage contains only
text in the proper format to correctly initialise the task array.

nguyiyang committed:
Improve code quality

Certain code is not of good quality

Better code quality can improve understandability.

Improving code quality throughout the codebase such as reducing deep
nesting, changing JavaDoc for string methods to provide clearer
understanding, abstract away code into method, refactor dialog box
into a function, etc.

Let's
* reduce deep nesting in code
* provide clearer comments
* apply SLAP
* minimize code duplication

nguyiyang committed:
Merge pull request #1 from nguyiyang/branch-A-Assertions

Add assertions to program

nguyiyang committed:
Resolve merge conflict with master branch

nguyiyang committed:
Merge pull request #2 from nguyiyang/branch-A-CodeQuality

Improve code quality

nguyiyang committed:
Add undo command

No undo command available for users.

Undo command provides convenience for users when making mistakes.

UndoCommand class is added to encapsulate actions related to reversing
the effects of add, delete and done commands, Log class is added to
keep track of such previous commands, Paser class is updated to support
undo commands.

Let's give users the option to undo all previous commands, by having
a Log class that can easily keep track of the history of commands
in which the UndoCommand class can easily use to reverse the actions.

nguyiyang committed:
Fix bug in UndoCommand

nguyiyang committed:
Merge branch 'branch-C-Undo'

nguyiyang committed:
Set up Continuous Integration

nguyiyang committed:
Remove I/O redirection tests

nguyiyang committed:
Fix CheckStyle errors

nguyiyang committed:
Add greeting message in GUI

nguyiyang committed:
Fix bug in undo delete command

nguyiyang committed:
Update GUI

nguyiyang committed:
Fix storage save bug

nguyiyang committed:
Fix undo command bugs for done commands

nguyiyang committed:
Update README.md

nguyiyang committed:
Set theme jekyll-theme-slate

nguyiyang committed:
Add tests for Command classes

nguyiyang committed:
Update coding standards and checkstyle

nguyiyang committed:
Merge branch 'master' of https://github.com/nguyiyang/ip
